src/System.Net.Sockets/src/System/Net/Sockets/OverlappedAsyncResult.Windows.cs

```diff
@@ -18,22 +18,21 @@ namespace System.Net.Sockets
     // from the BeginSend, BeginSendTo, BeginReceive, BeginReceiveFrom calls.
     internal partial class OverlappedAsyncResult : BaseOverlappedAsyncResult
     {
-        internal WSABuffer _singleBuffer;
-        internal WSABuffer[] _wsaBuffers;
-
-        internal IntPtr GetSocketAddressPtr()
+        internal unsafe byte* GetSocketAddressPtr()
         {
-            return Marshal.UnsafeAddrOfPinnedArrayElement(_socketAddress.Buffer, 0);
+            Debug.Assert(_socketAddress != null);
```

Comment:
And that Buffer is non-null as well?

(It's too bad there's no method like GC.IsPinned(object) we could assert on to verify that _socketAddress.Buffer is pinned, though I understand from @Maoni0 the runtime doesn't track pinning in a way that would allow that.)